martin/new-imports-ids

martinohansen opened the pull request:
Fix #35

hpernu commented:
Nice althoguh I don't quite follow what you're trying to achieve. Then again it is a draft. Do you even need comments yet?

martinohansen commented:
> Nice althoguh I don't quite follow what you're trying to achieve. Then again it is a draft. Do you even need comments yet?

Yep, its still work in progress. Feel free to comment on the general approach or come up with your own suggested implementation of a solution to this problem.

martinohansen commented on writer/ynab/ynab.go:
```diff
@@ -19,74 +18,115 @@ import (
-	if !found {
-		return fmt.Errorf("env variable YNAB_TOKEN: %w", ynabber.ErrNotFound)
+
+	// Version 2 method for generating YNAB import IDs. This is superior to v1
+	// because it uses the entire transaction to generate a hash.
+	v2 := func(y Ytransaction) (string, error) {
+		transaction, err := json.Marshal(y)
+		if err != nil {
+			return "", fmt.Errorf("failed to marshal YNAB transaction: %s: %w", y, err)
+		}
+		hash := sha256.Sum256(transaction)
+		return fmt.Sprintf("YBBR:%x", hash)[:32], nil
 	}
```
@hpernu: we might have a problem with using the entire transaction. If we do that any change to the transaction (memo, payee, etc.) will cause duplicate transactions in YNAB. We need to narrow it down to something that makes for better uniqueness while retaining some abstraction from the specifics of the transaction that might change.

hpernu commented:
Yes, I think the import ID or at least parts of it should be actually 
generated on Nordigen side
i.e. using parts of the structure that is read from Nordigen. This is 
more likely to remain constant for the same exact transaction and could 
be done using raw data. The translatons to payee/memo/whatever may 
change by user parameters or in the future because of other reasons. 
Using the original data should not change.

A particularly useful field here would be the Nordigen transaction id 
and IBAN of the account being read.

The way you're doing it now won't work. It will still result in same 
hash for a transaction on same date with same amount, to same payee, 
with NULL memo field. We for example have over 90% with NULL memo fields 
- practically always when the payment is to an external party. It is 
rarely useful as a distinguishing parameter although it won't hurt to 
include it.

I also think we can dispense with the YBBR: start allowing us to use as 
many characters of the hash as possible making conflicts less likely. Or 
is there a particular reason we want this? I understand if you want to 
export the data later from YNAB it would be easier to pinpoint ynabber 
generated transactions. Not a big deal, though, 32 characters is still 
plenty. But isn't the maximum length of import id 36 so the real 
effective length part here would be 31(not 32)?


On 12.11.2022 13.26, Martin Hansen wrote:
>
> ***@***.**** commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In writer/ynab/ynab.go 
> <https://github.com/martinohansen/ynabber/pull/36#discussion_r1020749254>:
>
> > +	// Version 2 method for generating YNAB import IDs. This is superior to v1
> +	// because it uses the entire transaction to generate a hash.
> +	v2 := func(y Ytransaction) (string, error) {
> +		transaction, err := json.Marshal(y)
> +		if err != nil {
> +			return "", fmt.Errorf("failed to marshal YNAB transaction: %s: %w", y, err)
> +		}
> +		hash := sha256.Sum256(transaction)
> +		return fmt.Sprintf("YBBR:%x", hash)[:32], nil
>   	}
>
> @hpernu <https://github.com/hpernu>: we might have a problem with 
> using the entire transaction. If we do that any change to the 
> transaction (memo, payee, etc.) will cause duplicate transactions in 
> YNAB. We need to narrow it down to something that makes for better 
> uniqueness while retaining some abstraction from the specifics of the 
> transaction that might change.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/martinohansen/ynabber/pull/36#pullrequestreview-1178114873>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ADNB667ZGHK4BPJ254GHQ4TWH55FFANCNFSM6AAAAAAR5SBK4U>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

martinohansen commented:
> Yes, I think the import ID or at least parts of it should be actually
> generated on Nordigen side
> i.e. using parts of the structure that is read from Nordigen. This is
> more likely to remain constant for the same exact transaction and could
> be done using raw data. The translatons to payee/memo/whatever may
> change by user parameters or in the future because of other reasons.
> Using the original data should not change.
> 
> A particularly useful field here would be the Nordigen transaction id
> and IBAN of the account being read.
> 
> The way you're doing it now won't work. It will still result in same
> hash for a transaction on same date with same amount, to same payee,
> with NULL memo field. We for example have over 90% with NULL memo fields
> -- practically always when the payment is to an external party. It is
> rarely useful as a distinguishing parameter although it won't hurt to
> include it.
> 
> I also think we can dispense with the YBBR: start allowing us to use as
> many characters of the hash as possible making conflicts less likely. Or
> is there a particular reason we want this? I understand if you want to
> export the data later from YNAB it would be easier to pinpoint ynabber
> generated transactions. Not a big deal, though, 32 characters is still
> plenty. But isn't the maximum length of import id 36 so the real
> effective length part here would be 31(not 32)?

I don't remember why I fixed it at 32 chars and I cant find it in the docs anymore. I think I added the YBBR because YNAB uses a similar prefix to their import IDs. 

I've refactored it to use the account ID and transaction ID. I'm going to do some further refactoring so the IBAN is used as account ID until we transform it for sending to YNAB. 

After that, all that is left -- as I see it -- is to write the ImportFromDate logic in the YNAB writer.

martinohansen commented:
@hpernu: a lot of refactoring but it was overdue. I tested this quickly and it does not break my existing config, so far so good. 

Before we merge this in we need to be somewhat certain we will not change the `ynabber.Account` type, and I'm not too certain of that currently. I would like to move the `nordigen.accountParser` function into the YNAB package and make it work on IBAN instead of Nordigen account IDs. In that way it will be more generic should we implement another writer down the line. 

What do you think?

hpernu commented:
> I don't remember why I fixed it at 32 chars and I cant find it in the docs anymore. I think I added the YBBR because > YNAB uses a similar prefix to their import IDs.

YNAB appears to use this prefix when importing from CSV.

Thus it would actually make more sense for as to use a different prefix or no prefix at all to distinguish from those imported from CSV. Perhaps just use _ so everything imported with ynabber would start with _ ?

I detected also that there is an upper limit of 36 characters for import id when using a different software for importing some other data to YNAB. Trying to import with a longer id will cause an error. I suggest adding a limiter in the same way we have for payee and memo which will cut the excess or otherwise making sure the length never exceeds that. In this case, this should be done silently though as the import id is internally generated by us.

hpernu commented:
A general comment: there are many changes, many of which are unrelated to import id generation. This makes reviewing them complicated. If you want me to properly review these, suggest splitting refactoring code to different pull requests/branches and moving completely unrelated things (such as panic with no transactions) to again different branches.

If we were to go through real acceptance cycle, there is now too many unrelated things, some of which I would not want to have (such as panic without no transactions) but the main issue(better import id generation) I do want. Now with all the unrelated and refactoring changes, it is difficult to follow the actual fairly simple change of import id generation logic. Refactoring is also best done separately where there is no new feature and nothing changes operationally so we can, via testing, be certain that everything still functions as before.

martinohansen commented:
> > I don't remember why I fixed it at 32 chars and I cant find it in the docs anymore. I think I added the YBBR because > YNAB uses a similar prefix to their import IDs.
> 
> YNAB appears to use this prefix when importing from CSV.
> 
> Thus it would actually make more sense for as to use a different prefix or no prefix at all to distinguish from those imported from CSV. Perhaps just use _ so everything imported with ynabber would start with _ ?
> 
> I detected also that there is an upper limit of 36 characters for import id when using a different software for importing some other data to YNAB. Trying to import with a longer id will cause an error. I suggest adding a limiter in the same way we have for payee and memo which will cut the excess or otherwise making sure the length never exceeds that. In this case, this should be done silently though as the import id is internally generated by us.

We are using a different prefix than YNAB, they use YNAB and we use YBBR. We already do limit it to 32 chars like with memo and payee: https://github.com/martinohansen/ynabber/blob/d58b85a55fbb3255372c20bf3530dd5f5d8f48e1/writer/ynab/ynab.go#L49-L53



> A general comment: there are many changes, many of which are unrelated to import id generation. This makes reviewing them complicated. If you want me to properly review these, suggest splitting refactoring code to different pull requests/branches and moving completely unrelated things (such as panic with no transactions) to again different branches.
> 
> If we were to go through real acceptance cycle, there is now too many unrelated things, some of which I would not want to have (such as panic without no transactions) but the main issue(better import id generation) I do want. Now with all the unrelated and refactoring changes, it is difficult to follow the actual fairly simple change of import id generation logic. Refactoring is also best done separately where there is no new feature and nothing changes operationally so we can, via testing, be certain that everything still functions as before.

Yes, I ended up going a bit wild with the refactoring, sorry about that. I did split everything into separate commits so you can review the refactoring and feature change separately and if needed we can also roll them back individually. 

I think you're misunderstanding the [panic commit](https://github.com/martinohansen/ynabber/pull/36/commits/d58b85a55fbb3255372c20bf3530dd5f5d8f48e1), I'm fixing a panic if all transactions are skipped, I'm **not** adding a panic one.

hpernu reviewed (changes requested):
Already asked to move refactoring to separate branch(out of this PR) so we only consider new id in here.

Also please consider instead of direct v1/v2 to just use v2 after a certain date but still use v1 for old data instead of completely skipping older data. At least in my case, I find it difficult to handle upgrade with this method.

An example:
- Assume I am changing method today
- I run ynabber once for all accounts to make sure all transactions are there (current v1 ids)
- I put all schedules on hold(this I have to do as I nowadays import automatically)
- I change the id generation method and put the date as today
- I rerun, already getting duplicates for today ... not good
   -> Or pehaps I put the date to tomorrow? ...
    Wait! What happens to possibly still missing transactions for today? Banks are not closed
     yet ... at least not digital transactions which may now happen at almost any time.
 -> Only certain way is to define a date after which new ids are used but use old ones for anything older than that.

In effect, I do not see a scenario where there is at least a small possibility of duplicates or missing some transactions. All of this is resolved if we do it like this:
- YNABBER_IMPORT_ID_V3 = 2099-12-01 (not really needed, just an example)
  If transaction date is lower, fall back to:
- YNABBER_IMPORT_ID_V2 = 2022-12-01
  If transaction date is lower, fall back to:
- YNABBER_IMPORT_ID_V1 = can't really set this but perhaps we can have special
   value never just to indicate to never use this.

The other option is also to LASTDATE parameter too. Then ynabber can be done with two configurations: LASTDATE set and using import id v1, or BEGINDATE (or whatever) set and import id v2.

hpernu commented:
Sorry, I am apparently saying approximately the same thing in multiple places. Still learning how this github PR-commenting should be used - more used to bitbucket as I predominantly work with closed source projects.

hpernu commented on config.go:
```diff
@@ -99,9 +99,29 @@ type YNAB struct {
+	V1 Date `envconfig:"YNAB_IMPORT_ID_V1" default:"1970-01-01"`
+
+	// V2 will be used from this date, for example: 2022-12-24
+	V2 Date `envconfig:"YNAB_IMPORT_ID_V2" default:"9999-01-01"`
```
Could we have the date by default as something like 2023-03-30 or some other date in the fairly near future? We could assume current users will upgrade eventually ...

martinohansen replied:
I don't see how we could assume that. I want to gather learnings about this feature before making any upgrade assumptions or cutover dates.

hpernu replied:
The best approach would be to actually use V2 imports by default for all new users and even for all new transactions. But I fail to see how this would be possible easily. The only way how this could be accomplished would be to query YNAB for existing transactions and see if there are any with an 'old' pattern. If none found use new, otherwise old ... but this becomes fairly complicated - I don't even know whether it is feasible at all. 

Perhaps, at some point, make these so that there are no defaults and the user must explicitly select this or the cutover date. We can even helpfully suggest if the parameter is missing on what to do. But this is another extra feature and will break current operation for existing users (although trivial to fix).

hpernu commented on writer/ynab/ynab.go:
```diff
@@ -36,7 +37,37 @@ type Ytransactions struct {
+	// Version 2 uses the account and transaction ID from ynabber.Transaction
+	v2Cutover := time.Time(cfg.YNAB.ImportID.V2)
+	v2 := func(t ynabber.Transaction) string {
+		hash := sha256.Sum256([]byte(t.Account.ID + t.ID))
```
Please add date and amount as part of the hash as well. They do not hurt, are certain to remain the same, and if, for whatever reason they were to ever change, this should result in a new transaction where the user will resolve this in YNAB. I don't think this kind of situation will ever be possible but we cannot know how different banks operate. Furthermore, we do not know how the transaction id is generated for different banks. Is it even unique ...

A specific example: with credit card transactions, there might be a much higher amount 'reserve' first which is later corrected with the right amount. I think in my case, the reservation is never imported, but we can't guarantee the operation is similar in all banks.

( To be honest, the particular case for reservation, if we see these, they should not be imported at all. But as it is, we do not know how to filter these out so this would have to be done if we get to know about these kinds of transactions appearing. )

martinohansen replied:
Done! 

Regarding the reserved transactions, I believe that they should be part of the "Pending" slice: https://github.com/frieser/nordigen-go-lib/blob/main/accounts.go#L76 and thus not be imported until they are "Booked": https://github.com/martinohansen/ynabber/blob/22906008c8606c77dfc94743a8344e92e8866c3f/reader/nordigen/nordigen.go#L109

hpernu replied:
Possible. In reality, we would really need extensive test data from multiple banks in different situations.

Perhaps we could, at some point, include a tool which will allow users to use ynabber to fetch Nordigen data, anonymize it (i.e. mangle all identification and private data) and sent it as test data.

hpernu reviewed (commented):
Made some comments but looks fine so far. I am otherwise ready to approve, but check the comments first.

Then a general question: this does not implement start date/end date feature but if it were to do so, wouldn't it make more sense to actually do that on the Nordigen(reader) side? That way we could reduce the size of request to Nordigen, and likely speed up the whole process and could reduce the number of transactions YNAB writer needs to process and YNAB needs to check for as duplicates. A separate feature, I know - perhaps I'll implement such a feature in the future.

martinohansen commented:
@hpernu: i think we're ready to get this merged in, do you wanna give it a last look?
